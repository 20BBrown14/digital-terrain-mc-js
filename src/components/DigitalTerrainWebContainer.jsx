import React from 'react';
import PropTypes from 'prop-types';
import selectedNavKeyPropTypes from '../proptypes/selectedNavKeyPropTypes';
import { NK_HOME } from '../constants/navKeys';
import DigitalTerrainWebView from './DigitalTerrainWebView';

const propTypes = {
  /* The page to load base on the requested url from server */
  selectedNavKey: selectedNavKeyPropTypes.isRequired,
  /* Authenitcated user's discord name */
  discordNick: PropTypes.string.isRequired,
  /* Whether authenticated user is admin */
  isAdmin: PropTypes.bool.isRequired,
  /* JWT Token */
  jwtToken: PropTypes.string.isRequired,
};

/**
 * Top level container. Handles nav management.
 */
class DigitalTerrainWebContainer extends React.Component {
  constructor(props) {
    super(props);
    const { selectedNavKey } = props;
    this.state = {
      selectedNavKey,
      homeCachedImages: [],
      galleryCachedImages: [],
    };

    this.handleNavMenuClick = this.handleNavMenuClick.bind(this);
    this.navigateToNewPage = this.navigateToNewPage.bind(this);
    this.setCachedHomeImages = this.setCachedHomeImages.bind(this);
    this.setCachedGalleryImages = this.setCachedGalleryImages.bind(this);
  }

  /**
   * Handles nav menu clicks. Sets selectedNavKey state
   * @param {Object} event - Click event information. See https://ant.design/components/menu/#API
   */
  handleNavMenuClick = (newNavKey) => {
    this.navigateToNewPage(newNavKey);
  };

  navigateToNewPage = (newNavKey) => {
    this.setState({ selectedNavKey: newNavKey === 'title' ? NK_HOME : newNavKey });
  }

  setCachedHomeImages = (imagesToCache) => {
    this.setState({ homeCachedImages: imagesToCache });
  }

  setCachedGalleryImages = (imagesToCache) => {
    this.setState({ galleryCachedImages: imagesToCache });
  }

  render() {
    const { selectedNavKey, homeCachedImages, galleryCachedImages } = this.state;
    const { discordNick, isAdmin, jwtToken } = this.props;
    return (
      <DigitalTerrainWebView
        selectedNavKey={selectedNavKey}
        handleNavMenuClick={this.handleNavMenuClick}
        navigateToNewPage={this.navigateToNewPage}
        homeCachedImages={homeCachedImages}
        setCachedHomeImages={this.setCachedHomeImages}
        galleryCachedImages={galleryCachedImages}
        setCachedGalleryImages={this.setCachedGalleryImages}
        discordNick={discordNick}
        isAdmin={isAdmin}
        jwtToken={jwtToken}
      />
    );
  }
}
DigitalTerrainWebContainer.propTypes = propTypes;

export default DigitalTerrainWebContainer;
