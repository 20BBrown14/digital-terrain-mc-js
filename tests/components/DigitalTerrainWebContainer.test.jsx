import React from 'react';
import {
  NK_HOME,
  NK_MAP,
} from '../../src/constants/navKeys';
import DigitalTerrainWebContainer from '../../src/components/DigitalTerrainWebContainer';

describe('DigitalTerrainWebContainer', () => {
  let container;
  beforeEach(() => {
    container = shallow(<DigitalTerrainWebContainer selectedNavKey={NK_HOME} />);
  });

  it('renders a default component', () => {
    container = render(<DigitalTerrainWebContainer selectedNavKey={NK_HOME} />);
    expect(container).toMatchSnapshot();
  });

  it('has a DigitalTerrainWebView', () => {
    const view = container.find('DigitalTerrainWebView');
    expect(view).toHaveLength(1);
  });

  it('sets default state', () => {
    expect(container.instance().state.selectedNavKey).toEqual(NK_HOME);
    expect(container.instance().state.homeCachedImages).toEqual([]);
    expect(container.instance().state.galleryCachedImages).toEqual([]);
  });

  describe('handleNavMenuClick', () => {
    let instance;
    beforeEach(() => {
      instance = container.instance();
    });

    it('calls navigation function', () => {
      instance.navigateToNewPage = jest.fn();
      instance.handleNavMenuClick(NK_MAP);
      expect(instance.navigateToNewPage).toHaveBeenCalledTimes(1);
      expect(instance.navigateToNewPage).toHaveBeenCalledWith(NK_MAP);
    });
  });

  describe('navigateToNewPage', () => {
    let instance;
    beforeEach(() => {
      instance = container.instance();
    });

    it('sets selectedNavKey to passed key', () => {
      expect(instance.state.selectedNavKey).toEqual(NK_HOME);
      instance.navigateToNewPage(NK_MAP);
      expect(instance.state.selectedNavKey).toEqual(NK_MAP);
    });

    it('sets selectedNavKey to home when passed title', () => {
      instance.state.selectedNavKey = NK_MAP;
      instance.navigateToNewPage('title');
      expect(instance.state.selectedNavKey).toEqual(NK_HOME);
    });
  });

  describe('setCachedHomeImages', () => {
    let instance;
    beforeEach(() => {
      instance = container.instance();
    });

    it('sets state as expected', () => {
      instance.setCachedHomeImages([{}, {}, {}, {}]);
      expect(instance.state.homeCachedImages).toEqual([{}, {}, {}, {}]);
    });
  });

  describe('setCachedGalleryImages', () => {
    let instance;
    beforeEach(() => {
      instance = container.instance();
    });

    it('sets state as expected', () => {
      instance.setCachedGalleryImages([{}, {}, {}, {}]);
      expect(instance.state.galleryCachedImages).toEqual([{}, {}, {}, {}]);
    });
  });
});
