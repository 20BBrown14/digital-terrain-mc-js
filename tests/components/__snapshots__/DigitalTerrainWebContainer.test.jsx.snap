// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DigitalTerrainWebContainer renders a default component 1`] = `
<section
  class="ant-layout web-view-layout"
>
  <header
    class="ant-layout-header web-view-layout-header"
  >
    <div>
      <ul
        class="ant-menu header-menu ant-menu-light ant-menu-root ant-menu-horizontal"
        role="menu"
      >
        <li
          class="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
          role="menuitem"
          style="display:none"
        >
          <div
            aria-expanded="false"
            aria-haspopup="true"
            class="ant-menu-submenu-title"
            role="button"
          >
            <span>
              ···
            </span>
            <i
              class="ant-menu-submenu-arrow"
            />
          </div>
        </li>
        <li
          aria-disabled="true"
          class="ant-menu-item header-icon-menu-item ant-menu-item-only-child ant-menu-item-disabled"
          role="menuitem"
        >
          <img
            alt="loading..."
            class="header-icon"
            src="test-file-stub"
          />
          <span />
        </li>
        <li
          class="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
          role="menuitem"
          style="display:none"
        >
          <div
            aria-expanded="false"
            aria-haspopup="true"
            class="ant-menu-submenu-title"
            role="button"
          >
            <span>
              ···
            </span>
            <i
              class="ant-menu-submenu-arrow"
            />
          </div>
        </li>
        <li
          class="ant-menu-item header-title header-non-highlighted-menu-item ant-menu-item-only-child"
          role="menuitem"
        >
          Digital Terrain
        </li>
        <li
          class="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
          role="menuitem"
          style="display:none"
        >
          <div
            aria-expanded="false"
            aria-haspopup="true"
            class="ant-menu-submenu-title"
            role="button"
          >
            <span>
              ···
            </span>
            <i
              class="ant-menu-submenu-arrow"
            />
          </div>
        </li>
        <li
          class="ant-menu-item ant-menu-item-only-child ant-menu-item-selected"
          role="menuitem"
        >
          Home
        </li>
        <li
          class="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
          role="menuitem"
          style="display:none"
        >
          <div
            aria-expanded="false"
            aria-haspopup="true"
            class="ant-menu-submenu-title"
            role="button"
          >
            <span>
              ···
            </span>
            <i
              class="ant-menu-submenu-arrow"
            />
          </div>
        </li>
        <li
          class="ant-menu-item ant-menu-item-only-child"
          role="menuitem"
        >
          Gallery
        </li>
        <li
          class="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
          role="menuitem"
          style="display:none"
        >
          <div
            aria-expanded="false"
            aria-haspopup="true"
            class="ant-menu-submenu-title"
            role="button"
          >
            <span>
              ···
            </span>
            <i
              class="ant-menu-submenu-arrow"
            />
          </div>
        </li>
        <li
          class="ant-menu-item ant-menu-item-only-child"
          role="menuitem"
        >
          Map
        </li>
        <li
          class="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
          role="menuitem"
          style="display:none"
        >
          <div
            aria-expanded="false"
            aria-haspopup="true"
            class="ant-menu-submenu-title"
            role="button"
          >
            <span>
              ···
            </span>
            <i
              class="ant-menu-submenu-arrow"
            />
          </div>
        </li>
        <li
          class="ant-menu-item ant-menu-item-only-child"
          role="menuitem"
        >
          Rules
        </li>
        <li
          class="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
          role="menuitem"
          style="display:none"
        >
          <div
            aria-expanded="false"
            aria-haspopup="true"
            class="ant-menu-submenu-title"
            role="button"
          >
            <span>
              ···
            </span>
            <i
              class="ant-menu-submenu-arrow"
            />
          </div>
        </li>
        <li
          class="ant-menu-item ant-menu-item-only-child"
          role="menuitem"
        >
          Server Information
        </li>
        <li
          class="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
          role="menuitem"
          style="display:none"
        >
          <div
            aria-expanded="false"
            aria-haspopup="true"
            class="ant-menu-submenu-title"
            role="button"
          >
            <span>
              ···
            </span>
            <i
              class="ant-menu-submenu-arrow"
            />
          </div>
        </li>
        <li
          class="ant-menu-item ant-menu-item-only-child"
          role="menuitem"
        >
          Apply
        </li>
        <li
          class="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
          role="menuitem"
          style="display:none"
        >
          <div
            aria-expanded="false"
            aria-haspopup="true"
            class="ant-menu-submenu-title"
            role="button"
          >
            <span>
              ···
            </span>
            <i
              class="ant-menu-submenu-arrow"
            />
          </div>
        </li>
        <li
          class="ant-menu-item ant-menu-item-only-child"
          role="menuitem"
        >
          About us
        </li>
        <li
          class="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
          role="menuitem"
          style="display:none"
        >
          <div
            aria-expanded="false"
            aria-haspopup="true"
            class="ant-menu-submenu-title"
            role="button"
          >
            <span>
              ···
            </span>
            <i
              class="ant-menu-submenu-arrow"
            />
          </div>
        </li>
        <li
          class="ant-menu-item ant-menu-item-only-child"
          role="menuitem"
        >
          <a
            href="https://discord.com/api/oauth2/authorize?client_id=693123471210709072&redirect_uri=http%3A%2F%2Flocalhost%3A3000%2Floggedin&response_type=code&scope=identify%20guilds"
          >
            Login
          </a>
        </li>
        <li
          class="ant-menu-submenu ant-menu-submenu-horizontal ant-menu-overflowed-submenu"
          role="menuitem"
          style="visibility:hidden;position:absolute"
        >
          <div
            aria-expanded="false"
            aria-haspopup="true"
            class="ant-menu-submenu-title"
            role="button"
          >
            <span>
              ···
            </span>
            <i
              class="ant-menu-submenu-arrow"
            />
          </div>
        </li>
      </ul>
    </div>
  </header>
  <main
    class="ant-layout-content web-view-layout-content"
  >
    <div
      class="home-page-content"
    >
      <h1
        class="ant-typography home-page-title"
      >
        Welcome to Digital Terrain
      </h1>
      <div
        class="ant-space ant-space-vertical"
      >
        <div
          class="ant-space-item"
          style="margin-bottom:8px"
        >
          <span
            class="ant-typography home-page-subtitle"
          >
            Digital Terrain is a SMP server inspired by Hermitcraft. The server is based on a great community, trust,          and fun. We aspire to have a playerbase of all play styles that work together in harmony to build great          things and have an awesome time. The server is owned by 
            <button
              class="home-page-link-button"
              id="pandaLink"
              type="button"
            >
              Panda_Brah
            </button>
             and 
            <button
              class="home-page-link-button"
              id="dmLink"
              type="button"
            >
              DMDF_
            </button>
            . Alongside them we have a team of mods who help make sure everything is running smoothly.
          </span>
        </div>
        <div
          class="ant-space-item"
        >
          <span
            class="ant-typography home-page-subtitle"
          >
            Be sure to checkout our 
            <button
              class="home-page-link-button"
              id="galleryLink"
              type="button"
            >
              gallery
            </button>
             and 
            <button
              class="home-page-link-button"
              id="mapLink"
              type="button"
            >
              map
            </button>
             to get a feel for the server; hopefully you'll find it's a good place for you!          If you decide to, you can 
            <button
              class="home-page-link-button"
              id="applyLink"
              type="button"
            >
              apply here
            </button>
             to be a part of our server. Hope to play with you soon!
          </span>
        </div>
      </div>
      <div
        class="ant-carousel"
      >
        <div
          class="slick-slider home-page-carousel slick-initialized"
        >
          <div
            class="slick-list"
          >
            <div
              class="slick-track"
              style="width:0%;left:NaN%"
            />
          </div>
        </div>
      </div>
    </div>
  </main>
  <footer
    class="ant-layout-footer"
  >
    © 2020 Copyright Digital Terrain
  </footer>
</section>
`;
